### PR TITLE
PWGHF: Add Topology Tag kinematic and topology table HFTag and probe

### PR DESCRIPTION
**Pull request by DanielMorris13**
This PR contains: 

- additional process function to collect tag topological and kinematic info in data and MC for DsPlus, Dstar
- helper function to collect tag pdg mother in MC and store topo. variables into the table

### CODE REVIEW

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx [hunk 1]
```diff
@@ -233,18 +262,103 @@ struct TagTwoProngDisplacedVertices {
...
+        }
+
+        ///  K∓K± for φ from Ds± or D± → φπ± decays
+        ///  K∓π± for D0 from D±* → D0π± decays
+        else {
+          for (auto pdgGrandMother : pdgDecayMothers) {
+            auto grandMotherId = RecoDecay::getMother(particlesMc, particleMother, pdgGrandMother, true);
+            auto particleGrandMother = particlesMc.rawIteratorAt(grandMotherId);
+            if (grandMotherId != -1) {
+              auto daughters = particleGrandMother.template daughters_as<aod::McParticles>();
```
Comment (mfaggin) [inline DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:339]: same as above

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx [hunk 2]
```diff
@@ -233,18 +262,103 @@ struct TagTwoProngDisplacedVertices {
...
+            }
+          }
+        }
+
+        ///  K∓K± for φ from Ds± or D± → φπ± decays
+        ///  K∓π± for D0 from D±* → D0π± decays
+        else {
+          for (auto pdgGrandMother : pdgDecayMothers) {
+            auto grandMotherId = RecoDecay::getMother(particlesMc, particleMother, pdgGrandMother, true);
+            auto particleGrandMother = particlesMc.rawIteratorAt(grandMotherId);
```
Comment (mfaggin) [inline DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:337]: what about moving it inside the `if (grandMotherId != -1) {` statement?

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx [hunk 3]
```diff
@@ -233,18 +262,103 @@ struct TagTwoProngDisplacedVertices {
...
+                if (std::abs(daughter.pdgCode()) == pdgProbeParticle) {
+                  return 1;
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    return 0;
```
Comment (mfaggin) [inline DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:351]: here you return `0`, while above when either one of the daughter does not have a MC particle you return `-999`. Is it what you want? do you really want to separate the two cases with the return? Nothing against it, I just raise the point in case you did not ask it yourself.

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx [hunk 4]
```diff
@@ -233,18 +262,103 @@ struct TagTwoProngDisplacedVertices {
...
+  }
+
   template <typename Pvec>
   bool isSelectedInvariantMass(const Pvec& pVecTrackFirst,
                                const Pvec& pVecTrackSecond,
                                const uint8_t channel,
-                               float& invMass2)
+                               float& invMass2,
+                               const int& ptBin)
```
Comment (mfaggin) [inline DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:359]: why did you add this? where do you use this `ptBin` in the function?

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx [hunk 5]
```diff
@@ -299,14 +459,9 @@ struct TagTwoProngDisplacedVertices {
                           const CovMatSV& covMatrixSecVtx,
                           const PVec& pVec,
                           std::array<float, 2>& trackDcaXy,
-                          const uint8_t channel)
+                          const uint8_t channel,
+                          const int& ptBin)
```
Comment (mfaggin) [inline DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:463]: also here I do not understand where you use this new argument

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx [hunk 6]
```diff
@@ -372,7 +529,12 @@ struct TagTwoProngDisplacedVertices {
         std::array<float, 3> pVecTrackFirst{trackFirst.px(), trackFirst.py(), trackFirst.pz()};
         std::array<float, 3> pVecTrackSecond{trackSecond.px(), trackSecond.py(), trackSecond.pz()};
 
-        if (!isSelectedInvariantMass(pVecTrackFirst, pVecTrackSecond, channel, invMass2)) {
+        auto pVec = RecoDecay::pVec(pVecTrackFirst, pVecTrackSecond);
+        auto ptBin = findBin(&ptBinsForTopologicalCuts[channel], RecoDecay::pt(pVec));
+        if (ptBin == -1)
+          continue;
+
```
Comment (mfaggin) [inline DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:536]:
```suggestion
        if (ptBin == -1) {
          continue;
        }

```

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx [hunk 7]
```diff
@@ -427,7 +594,13 @@ struct TagTwoProngDisplacedVertices {
         float invMass2{0.f};
         std::array<float, 3> pVecTrackPos{trackPos.px(), trackPos.py(), trackPos.pz()};
         std::array<float, 3> pVecTrackNeg{trackNeg.px(), trackNeg.py(), trackNeg.pz()};
-        if (!isSelectedInvariantMass(pVecTrackPos, pVecTrackNeg, channel, invMass2)) {
+
+        auto pVec = RecoDecay::pVec(pVecTrackPos, pVecTrackNeg);
+        auto ptBin = findBin(&ptBinsForTopologicalCuts[channel], RecoDecay::pt(pVec));
+        if (ptBin == -1)
+          continue;
+
```
Comment (mfaggin) [inline DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:602]:
```suggestion
        if (ptBin == -1) {
          continue;
        }

```

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx [hunk 8]
```diff
@@ -445,65 +618,97 @@ struct TagTwoProngDisplacedVertices {
...
 
-        auto pVec = RecoDecay::pVec(pVecTrackPos, pVecTrackNeg);
         auto primVtx = getPrimaryVertex(collision);
         const auto& secVtx = vertexer.getPCACandidate();
         const auto& covMatrixPCA = vertexer.calcPCACovMatrixFlat();
         std::array<float, 2> trackDcaXy{trackPos.dcaXY(), trackNeg.dcaXY()};
-        if (!isSelectedTopology(primVtx, secVtx, covMatrixPCA, pVec, trackDcaXy, channel)) {
+        if (fillTagTable) {
+          getTagInfo<doMc>(primVtx, secVtx, covMatrixPCA, pVec, trackDcaXy, channel, trackPos, trackNeg, invMass2, pdgDecayMothers, particlesMc);
           continue;
```
Comment (mfaggin) [inline DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:627]: do you really need it? Aren't you at the end of the iteration?

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx [hunk 9]
```diff
@@ -390,27 +552,32 @@ struct TagTwoProngDisplacedVertices {
...
 
-        auto pVec = RecoDecay::pVec(pVecTrackFirst, pVecTrackSecond);
         auto primVtx = getPrimaryVertex(collision);
         const auto& secVtx = vertexer.getPCACandidate();
         const auto& covMatrixPCA = vertexer.calcPCACovMatrixFlat();
         std::array<float, 2> trackDcaXy{trackFirst.dcaXY(), trackSecond.dcaXY()};
-        if (!isSelectedTopology(primVtx, secVtx, covMatrixPCA, pVec, trackDcaXy, channel)) {
+        if (fillTagTable) {
+          getTagInfo<doMc>(primVtx, secVtx, covMatrixPCA, pVec, trackDcaXy, channel, trackFirst, trackSecond, invMass2, pdgDecayMothers, particlesMc);
           continue;
```
Comment (mfaggin) [inline DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:561]: do you really need it? Aren't you at the end of the iteration?

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx [hunk 10]
```diff
@@ -233,18 +262,103 @@ struct TagTwoProngDisplacedVertices {
...
+            if (std::abs(daughter.pdgCode()) == pdgProbeParticle) {
+              return 1;
+            }
+          }
+        }
+
+        ///  K∓K± for φ from Ds± or D± → φπ± decays
+        ///  K∓π± for D0 from D±* → D0π± decays
+        else {
+          for (auto pdgGrandMother : pdgDecayMothers) {
```
Comment (mfaggin) [inline DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:335]: in case you use `DsOrDplusToKKPi` you return 1 either if the kaon has a `Ds` or a `D+` as mother, but in this way is not distinguishable. Is it what you want?

**Review by mfaggin — CHANGES_REQUESTED** [review]
thanks Daniel for the work! Please find here my suggestions

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx (hunk 10 above)
Comment (DanielMorris13) [reply DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:335]: Thank you for the comment! 
For the time being we are following the Fabrizio Grosa recipe where he does not  make any selection on the mother for this particular decay channel. Here we are just looking that both Kaons come from the Phi.
https://indico.cern.ch/event/1371921/contributions/5773861/attachments/2787104/4859629/AOTTrack_Dmeson_TagAndProbe.pdf

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx (hunk 8 above)
Comment (DanielMorris13) [reply DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:627]: You are right. 
I've implemented the correction.

Comment (DanielMorris13) [reply DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:627]: You are right. 
I've implemented the correction.

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx (hunk 9 above)
Comment (DanielMorris13) [reply DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:561]: Same as above.

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx [hunk 11]
```diff
@@ -233,18 +262,103 @@ struct TagTwoProngDisplacedVertices {
...
+      pdgTagMother = constants::physics::Pdg::kD0;
+      pdgProbeParticle = 211; // Pi
+    }
+
+    if (!firsTrack.has_mcParticle() || !secondTrack.has_mcParticle()) {
+      return -999;
+    } else {
+      auto firstMcTrack = firsTrack.template mcParticle_as<aod::McParticles>();
+      auto secondMcTrack = secondTrack.template mcParticle_as<aod::McParticles>();
+      auto firstTrackMotherId = RecoDecay::getMother(particlesMc, firstMcTrack, pdgTagMother, true);
```
[the remark this answers is not in the document]
Comment (DanielMorris13) [reply DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:314]: Thank you! I've implemented the correction.

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx (hunk 3 above)
Comment (DanielMorris13) [reply DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:351]: You are right. We don't need this kind of further distinguish. Thank you!

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx (hunk 4 above)
Comment (DanielMorris13) [reply DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:359]: In line 533 and 599 we calculate the ptBin once for each function "computeCombinatorialSameCharge" and "computeCombinatorialOppositeCharge". Then we pass this information to the functions which need it. Particularly, in this case we use this variable to get the topological cuts of interest.
Thank you.

File: DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx [hunk 12]
```diff
@@ -121,8 +143,9 @@ struct TagTwoProngDisplacedVertices {
   Produces<aod::KaKaFromDspTags> tagKaKaTable;
   Produces<aod::KaPiFromDzTags> tagKaPiTable;
   Produces<aod::PiKaFromDzTags> tagPiKaTable;
+  Produces<aod::TagTopoVariables> tagVarsTable;
   SliceCache cache;
-
+  Configurable<bool> fillTagTable{"fillTagTable", true, "flag to fill tag table with topological variables"};
```
[the remark this answers is not in the document]
Comment (DanielMorris13) [reply DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx:148]: Thank you. You are right!

**Comment by DanielMorris13** [comment]
Hi Mattia!
Thank you for the comments. I implemented the correction.

Daniel.

**Comment by mfaggin** [comment]
> Hi Mattia! Thank you for the comments. I implemented the correction.
> 
> Daniel.

Thanks Daniel, once you fix the formatting issues I approve it

**Review by mfaggin — DISMISSED** [review]

**Review by mfaggin — APPROVED** [review]
